### PR TITLE
[SPARK-13223] [ML] Add stratified sampling to ML feature engineering

### PR DESCRIPTION
**Pull request by hhbyyh**
jira: https://issues.apache.org/jira/browse/SPARK-13223

I found it useful to add an sampling transformer during a case of fraud detection. It can be used in resampling or overSampling, which in turn is required by ensemble and unbalanced data processing.
Internally, it invoke the sampleByKey in Pair RDD operation.

Ideally, it should take the fraction parameter as Map[Any, Double], yet that would bring difficulty in model peresistence. Appreciate if anyone can show how to save Map[Any, Double] to a DataFrame.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.feature.StratifiedSampling.StratifiedSamplingWriter
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: Import this inside the class.

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+ * :: Experimental ::
+ *
+ * Stratified sampling on the DataFrame according to the keys in a specific label column. User
+ * can set 'fraction' to set different sampling rate for each key.
+ *
+ * @param withReplacement can elements be sampled multiple times (replaced when sampled out)
+ * @param fraction expected size of the sample as a fraction of the items
+ *  without replacement: probability that each element is chosen; fraction must be [0, 1]
+ *  with replacement: expected number of times each element is chosen; fraction must be >= 0
+ */
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: Add `@see` to link to `DataFrame.sampleBy` so we don't need to document the behavior for keys not appearing in the map.

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+@Experimental
+final class StratifiedSampling private(
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: * `StratifiedSampler`
* space after `private`

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    override val uid: String,
+    val withReplacement: Boolean,
+    val fraction: Map[String, Double])
+  extends Transformer with HasLabelCol with HasSeed with DefaultParamsWritable {
+
+  @Since("2.0.0")
+  def this(withReplacement: Boolean, fraction: Map[String, Double]) =
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: Add a Java friendly constructor. `Map` is from Scala.

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setSeed(value: Long): this.type = set(seed, value)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setLabel(value: String): this.type = set(labelCol, value)
+
+  setDefault(seed -> Utils.random.nextLong)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: This is not necessary. We assign a fixed seed based on the hash value of the class name.

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+  @Since("2.0.0")
+  override def transform(data: DataFrame): DataFrame = {
+    transformSchema(data.schema)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: turn on logging

File: mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    val schema = data.schema
+    val colId = schema.fieldIndex($(labelCol))
+    val result = data.rdd.map(r => (r.get(colId), r))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala]: `DataFrame` has `sampleBy` implemented. We can just use that.

**Comment by mengxr** [comment]
@hhbyyh I think this is good to add. However, we need to think about the behavior during prediction. Certainly we don't want to apply sampling during prediction, but this is not addressed by the current pipeline implementation. Not for this PR, I just want to collect some ideas about this issue.

**Comment by hhbyyh** [comment]
@mengxr Thanks for the review. Sorry for the late response, I was on a flight.

It's great to know DataFrame.stat.sampleB. One concern is that it does not allow `withReplacement = true`. That means oversampling is not supported.

For prediction, are you worrying about that users need to use the same PipelineModel for training and evaluation dataset? I would propose the solution to explicitly allow stepwise enable/disable on each stage for a PipelineModel. Thus users can skip specific steps in a pipeline.

**Comment by hhbyyh** [comment]
@mengxr Appreciate if you have time for this

**Comment by hhbyyh** [comment]
I tried to make it support multiple data types. As shown in the code, the main constraint is the save/load implementation. It'll be great if we can support Any key in the save/load.

**Comment by hhbyyh** [comment]
Close it since it's been overlooked for some time and can be implemented with https://github.com/apache/spark/pull/17583 easily. Thanks for the review and comments.

**Pull request closed without merging**